### PR TITLE
Compile the react bundle with GCC advanced mode

### PR DESCRIPTION
**Pull request by banga**
Follow-up from #11967 to only compile the `react` bundle with GCC advanced mode. I also cleaned up the externs here to be consistent with the current public API. I also removed the type annotations from the externs file, since we don't use GCC for type-checking anyway.

### CODE REVIEW

File: externs/react.ext.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+React.Component.prototype.shouldComponentUpdate;
+React.Component.prototype.componentWillUpdate;
+React.Component.prototype.componentDidUpdate;
+React.Component.prototype.componentWillUnmount;
+React.Component.prototype.componentDidCatch;
+React.Component.prototype.render;
+
+React.PureComponent.prototype.isPureReactComponent;
+
+React.SyntheticEvent;
```
Comment (gaearon) [inline externs/react.ext.js]: I wouldn't expect this to be in public API of React package

File: externs/react.ext.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+React.SyntheticEvent.prototype.dispatchConfig;
+React.SyntheticEvent.prototype.nativeEvent;
+React.SyntheticEvent.prototype._targetInst;
+React.SyntheticEvent.prototype.isDefaultPrevented;
+React.SyntheticEvent.prototype.isPropagationStopped;
+React.SyntheticEvent.prototype._dispatchListeners;
+React.SyntheticEvent.prototype._dispatchInstances;
+
+/** @suppress {duplicate} */
+var DispatchConfig;
```
Comment (gaearon) [inline externs/react.ext.js]: Same

File: externs/react.ext.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+DispatchConfig.dependencies;
+DispatchConfig.phasedRegistrationNames = {};
+DispatchConfig.phasedRegistrationNames.bubbled;
+DispatchConfig.phasedRegistrationNames.captured;
+DispatchConfig.registrationName;
+
+/** @suppress {duplicate} */
+var FiberNode = {};
```
Comment (gaearon) [inline externs/react.ext.js]: Same, this shouldn't be in React package

File: externs/react.ext.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+FiberNode.prototype.lastEffect;
+FiberNode.prototype.expirationTime;
+FiberNode.prototype.alternate;
+FiberNode.prototype._debugID;
+FiberNode.prototype._debugSource;
+FiberNode.prototype._debugOwner;
+FiberNode.prototype._debugIsCurrentlyTiming;
+
+var Deadline = {};
+Deadline.prototype.timeRemaining;
```
Comment (gaearon) [inline externs/react.ext.js]: Same

File: externs/react.ext.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+var Reconciler = {};
```
Comment (gaearon) [inline externs/react.ext.js]: Same

File: externs/react.ext.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+var type;
+var key;
+var ref;
+var props;
+var _owner;
+
+React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = {};
+React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.assign;
+React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner;
+React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.EventPluginHub;
```
Comment (gaearon) [inline externs/react.ext.js]: This shouldn't be here (same for all of them below). It's on ReactDOM's field, not React's.

File: externs/fbjs.ext.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/** @suppress {duplicate} */
```
Comment (koba04) [inline externs/fbjs.ext.js]: Why is this file required?

Comment (banga) [reply externs/fbjs.ext.js]: This is only needed for the node bundles because the contents of fbjs are not included in those bundles, so properties like `emptyFunction.thatReturns` get renamed when they shouldn't be. I also remember hitting a problem with `canUseDOM` not being set to true when it should be, but I can't repro that right now. I was being defensive when I created this file by including everything from fbjs that I saw being used, but we can cut it down to a smaller set of things if that would be preferable.

Comment (banga) [reply externs/fbjs.ext.js]: Forgot to mention — this file is not pulled in for UMD bundles, because fbjs is not part of `externals` for those, which are checked in `getClosureExterns` to determine which externs files to look for.

Comment (koba04) [reply externs/fbjs.ext.js]: I got it. Thank you!

**Comment by banga** [comment]
Hey @gaearon, just checking if you got a chance to look at this.

**Comment by banga** [comment]
(Whoops, tapped the wrong button)

**Comment by gaearon** [comment]
Hey @banga,

This is very valuable but my focus has shifted from working on the repo infrastructure to React app compilation experiments so I won't have time to fully review/merge/support this in the near future.

I think it's safe to assume that whenever we decide to get back to this, the changes won't be very different (aside from fixing up the whitelists). I'll happily do that once I feel comfortable with spending more time on this. Please be assured this work was not in vain. This is the first time we actually had a complete PR that implements this, and it shows exactly the parts we need to fix first. (e.g. I'm not very happy with whitelists like internal variable names having to be whitelisted—this means we need to rewrite those parts in a way that doesn't rely on those names.)

Does this make sense? I think we can revisit in a few months.

**Comment by banga** [comment]
Yeah, that makes sense. Completely agree about rewriting the parts that share internal names. Ideally the whitelist should exactly match the public react API.
Whenever you get back to this, another thing that might be worth thinking about is explicitly supporting compiling React with other codebases in advanced mode. That’s one place where most of the externs (except for some DOM and css property names) are not needed, so you get the most wins. We do this at @quip, but we end up having to fix the things that break in advanced mode, which also makes version updates painful.
Thanks!

File: externs/react.ext.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
+/** @suppress {duplicate} */
```
Comment (phistuck) [inline externs/react.ext.js]: While there is some value in having a clearly defined API somewhere (which can also be used as an externs file for other project, even better if it had type annotations), why are you not using the `@expose` annotation instead?

File: packages/react/src/ReactElement.js [hunk 9]
```diff
@@ -10,20 +10,15 @@ import {REACT_ELEMENT_TYPE} from 'shared/ReactSymbols';
 
 import ReactCurrentOwner from './ReactCurrentOwner';
 
-const hasOwnProperty = Object.prototype.hasOwnProperty;
+const objHasOwnProperty = Object.prototype.hasOwnProperty;
 
-const RESERVED_PROPS = {
```
Comment (Deraen) [inline packages/react/src/ReactElement.js]: Related to #12368 #12370 #12530

This RESERVED_PROPS change longer required because Closure-compiler was fixed to not collapse properties from objects which are accessed using `hasOwnProperty`: https://github.com/google/closure-compiler/commit/c13cf48b98477e44409dba6359246bffa95b1c7b

This fix is included in release v20180319

Based on discussions on the mentioned PR's I think the change should be reverted.

**Comment by epicfaace** [comment]
Any plans to revisit this any time soon?

**Comment by darkleaf** [comment]
+
